Adding python based e2e tests for apigatewayv2 controller.

vijtrip2 opened the pull request:
Description of changes:
* Adding python based e2e tests for apigatewayv2 controller.

Test output:

```
➜  aws-controllers-k8s git:(pytest) ✗ make kind-test
checking AWS credentials ... ok.
creating kind cluster ack-test-0e3baec6-5653727a ... ok.
[+] Building 370.1s (21/21) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                                           0.1s
 => => transferring dockerfile: 39B                                                                                                                                                                                                            0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                              0.1s
 => => transferring context: 2B                                                                                                                                                                                                                0.0s
 => [internal] load metadata for docker.io/library/amazonlinux:2                                                                                                                                                                               3.6s
 => [internal] load metadata for docker.io/library/golang:1.14.1                                                                                                                                                                               3.6s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                                                                                                                                  0.0s
 => [auth] library/amazonlinux:pull token for registry-1.docker.io                                                                                                                                                                             0.0s
 => [builder  1/11] FROM docker.io/library/golang:1.14.1@sha256:38a520b05b971181a5a30d72b74d1b70c63688eb267a4fb376c7476a5c6f885b                                                                                                               0.2s
 => => resolve docker.io/library/golang:1.14.1@sha256:38a520b05b971181a5a30d72b74d1b70c63688eb267a4fb376c7476a5c6f885b                                                                                                                         0.2s
 => [internal] load build context                                                                                                                                                                                                              0.5s
 => => transferring context: 8.71kB                                                                                                                                                                                                            0.3s
 => CACHED [stage-1 1/3] FROM docker.io/library/amazonlinux:2@sha256:06380711d6a8ac0b6989f7e2a4419e560796791d9c7c843753a719c73552dc30                                                                                                          0.0s
 => CACHED [builder  2/11] WORKDIR /github.com/aws-controllers-k8s/apigatewayv2-controller                                                                                                                                                     0.0s
 => CACHED [builder  3/11] COPY apigatewayv2-controller/LICENSE /github.com/aws-controllers-k8s/apigatewayv2-controller/LICENSE                                                                                                                0.0s
 => CACHED [builder  4/11] COPY apigatewayv2-controller/ATTRIBUTION.md /github.com/aws-controllers-k8s/apigatewayv2-controller/ATTRIBUTION.md                                                                                                  0.0s
 => CACHED [builder  5/11] COPY apigatewayv2-controller/go.mod /github.com/aws-controllers-k8s/apigatewayv2-controller/go.mod                                                                                                                  0.0s
 => CACHED [builder  6/11] COPY apigatewayv2-controller/go.sum /github.com/aws-controllers-k8s/apigatewayv2-controller/go.sum                                                                                                                  0.0s
 => [builder  7/11] RUN go mod download                                                                                                                                                                                                       76.0s
 => [builder  8/11] COPY apigatewayv2-controller/apis /github.com/aws-controllers-k8s/apigatewayv2-controller/apis                                                                                                                             0.2s
 => [builder  9/11] COPY apigatewayv2-controller/cmd /github.com/aws-controllers-k8s/apigatewayv2-controller/cmd                                                                                                                               0.1s
 => [builder 10/11] COPY apigatewayv2-controller/pkg /github.com/aws-controllers-k8s/apigatewayv2-controller/pkg                                                                                                                               0.2s
 => [builder 11/11] RUN GIT_VERSION=30ecf27 &&     GIT_COMMIT=30ecf27019adce0ced02728e8dd907f15741af2b &&     BUILD_DATE=2021-04-04T19:36 &&     go build -ldflags="-X github.com/aws-controllers-k8s/apigatewayv2-controller/pkg/version.G  286.1s
 => [stage-1 2/3] COPY --from=builder /github.com/aws-controllers-k8s/apigatewayv2-controller/bin/controller /github.com/aws-controllers-k8s/apigatewayv2-controller/LICENSE /github.com/aws-controllers-k8s/apigatewayv2-controller/ATTRIBUT  0.8s
 => exporting to image                                                                                                                                                                                                                         0.5s
 => => exporting layers                                                                                                                                                                                                                        0.5s
 => => writing image sha256:82c13778d1926f24e968afc18a33a9f31a8c2c4576afa2971f5872fc64bcacb8                                                                                                                                                   0.0s
 => => naming to docker.io/library/aws-controllers-k8s:apigatewayv2-v0.0.2-69-gc071d23-dirty                                                                                                                                                   0.0s
ok.
loading the images into the cluster ... ok.
loading CRD manifests for apigatewayv2 into the cluster ... ok.
loading RBAC manifests for apigatewayv2 into the cluster ... ok.
loading service controller Deployment for apigatewayv2 into the cluster ...ok.
generating AWS temporary credentials and adding to env vars map ... ok.
======================================================================================================
To poke around your test cluster manually:
export KUBECONFIG=/Users/vijat/Documents/gocode/src/github.com/vijtrip2/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-0e3baec6-5653727a/kubeconfig
kubectl get pods -A
======================================================================================================
Adding python based e2e tests for apigatewayv2 controller.
testing Helm release for apigatewayv2 for release version v0.0.2-69-gc071d23-dirty.
Building release artifacts for apigatewayv2-v0.0.2-69-gc071d23-dirty
Generating custom resource definitions for apigatewayv2
Generating RBAC manifests for apigatewayv2
namespace/ack-system-test-helm created
installing the helm chart for ack-apigatewayv2-controller in namespace ack-system-test-helm ... ok.
uninstalling the helm chart for ack-apigatewayv2-controller in namespace ack-system-test-helm ... ok.
namespace "ack-system-test-helm" deleted
INFO:root:Wrote bootstrap to /root/tests/apigatewayv2/bootstrap.yaml
=============================================================================================================== test session starts ================================================================================================================
platform linux -- Python 3.8.8, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /root/tests
plugins: xdist-2.2.0, ordering-0.6, forked-1.3.0
[gw1] Python 3.8.8 (default, Feb 19 2021, 18:07:06)  -- [GCC 8.3.0]
[gw0] Python 3.8.8 (default, Feb 19 2021, 18:07:06)  -- [GCC 8.3.0]
[gw2] Python 3.8.8 (default, Feb 19 2021, 18:07:06)  -- [GCC 8.3.0]
gw0 [7] / gw1 [7] / gw2 [7]
scheduling tests via LoadFileScheduling

apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_perform_invocation
[gw1] [ 14%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_perform_invocation
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_httpapi
[gw1] [ 28%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_httpapi
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_httpapi_using_import
[gw1] [ 42%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_httpapi_using_import
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_integration
[gw1] [ 57%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_integration
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_authorizer
[gw1] [ 71%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_authorizer
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_route
[gw1] [ 85%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_route
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_stage
[gw1] [100%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_stage

========================================================================================================== 7 passed in 359.85s (0:05:59) ===========================================================================================================
To resume test with the same cluster use: " TMP_DIR=/Users/vijat/Documents/gocode/src/github.com/vijtrip2/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-0e3baec6-5653727a
    AWS_SERVICE_DOCKER_IMG=aws-controllers-k8s:apigatewayv2-v0.0.2-69-gc071d23-dirty "
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

RedbackThomson commented on test/e2e/apigatewayv2/service_bootstrap.py:
```diff
@@ -0,0 +1,95 @@
+    except lambda_client.exceptions.ResourceNotFoundException:
+        pass
+
+    current_directory = os.path.dirname(os.path.realpath(__file__))
+    index_zip = ZipFile(f'{current_directory}/resources/index.zip', 'w')
```
nit: Use a temp dir (https://docs.python.org/3/library/tempfile.html)

jaypipes replied:
Yes, this is a little more than a nit, though. :) Definitely should be using `tempfile`

vijtrip2 replied:
Done.

jaypipes replied:
I don't see a use of tempfile?

RedbackThomson commented on test/e2e/apigatewayv2/service_bootstrap.py:
```diff
@@ -0,0 +1,95 @@
+        pass
+
+    resp = iam_client.create_role(
+        RoleName=AUTHORIZER_IAM_ROLE_NAME,
```
I'm a little concerned about reusing the same resource names. In the SageMaker tests create a new role, with a randomised suffix, every time we bootstrap. This ensures we don't accidentally override another run (if running in parallel), or that tests don't fail when creating a resource that already exists.

vijtrip2 replied:
Very good suggestion, I will update it.

RedbackThomson commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+def apigw_validator():
+    return ApiGatewayValidator(boto3.client('apigatewayv2'))
+
+@pytest.fixture(scope="module")
+def test_resource_values():
+    return REPLACEMENT_VALUES.copy()
```
These don't need to be a fixture, you can just use them as globally scoped variables within tests?

vijtrip2 replied:
Done.

RedbackThomson reviewed (commented):
Well structured tests, seems logically laid out and your helper methods are a nice touch.

I like your use of `pytest-ordering`, although with some of the suggestions Jay had for the SageMaker integration tests (https://github.com/aws-controllers-k8s/community/pull/713#discussion_r592441442), this could be replaced with a simpler model.

jaypipes commented on test/e2e/apigatewayv2/resources/import_api.yaml:
```diff
@@ -0,0 +1,25 @@
+                    "uri": "http://example.com",
+                    "httpMethod": "GET",
+                    "type": "HTTP_PROXY",
+                    "payloadFormatVersion": "1.0"
```
Not sure if it matters, but the payloadFormatVersion for the authorizer above is 2.0, and this is 1.0?

vijtrip2 replied:
It does not matter. 
For integrations, only supported version is 1.0 and this field was introduced to evolve in future.
However for authorizers format version "2.0" is simplified response format and we use it in tests. : )

jaypipes commented on test/e2e/apigatewayv2/tests/helper.py:
```diff
@@ -0,0 +1,144 @@
+    return cr['status']['apiEndpoint']
+
+
+class ApiGatewayValidator:
```
nice separation of assertions here.

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+@pytest.mark.canary
+class TestApiGatewayV2:
+
+    @pytest.mark.run(order=1)
+    def test_precheck_api_resource(self, test_resource_values):
+        api_resource_name = test_resource_values['API_NAME']
+        if k8s.get_resource_exists(helper.get_api_ref(api_resource_name)):
+            pytest.fail(f"expected {api_resource_name} to not exist. Did previous test cleanup?", False)
+
+    @pytest.mark.run(order=2)
+    def test_precheck_integration_resource(self, test_resource_values):
+        integration_resource_name = test_resource_values['INTEGRATION_NAME']
+        if k8s.get_resource_exists(helper.get_api_ref(integration_resource_name)):
+            pytest.fail(f"expected {integration_resource_name} to not exist. Did previous test cleanup?", False)
+
+    @pytest.mark.run(order=3)
+    def test_precheck_authorizer_resource(self, test_resource_values):
+        authorizer_resource_name = test_resource_values['AUTHORIZER_NAME']
+        if k8s.get_resource_exists(helper.get_api_ref(authorizer_resource_name)):
+            pytest.fail(f"expected {authorizer_resource_name} to not exist. Did previous test cleanup?", False)
+
+    @pytest.mark.run(order=4)
+    def test_precheck_route_resource(self, test_resource_values):
+        route_resource_name = test_resource_values['ROUTE_NAME']
+        if k8s.get_resource_exists(helper.get_api_ref(route_resource_name)):
+            pytest.fail(f"expected {route_resource_name} to not exist. Did previous test cleanup?", False)
+
```
Recommend replacing these "precheck" methods with pytest fixtures eventually.

vijtrip2 replied:
Done.

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+
+    @pytest.mark.run(order=8)
+    def test_create_authorizer(self, apigw_validator, test_resource_values):
+        # Create an integration for the HTTP Api
```
s/integration/authorizer/

vijtrip2 replied:
Done.

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+            "authorizer",
+            additional_replacements=test_resource_values
+        )
+        logging.error(f"authorizer resource: {authorizer_resource_data}")
```
This should not be `logging.error`...

vijtrip2 replied:
Changed to info.

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+            "integration",
+            additional_replacements=test_resource_values
+        )
+        logging.error(f"integration resource: {integration_resource_data}")
```
This should not be `logging.error`...

vijtrip2 replied:
changed to info

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+        )
+        logging.error(f"authorizer resource: {authorizer_resource_data}")
+
+        # Create the k8s integration resource
```
s/integration/authorizer/

vijtrip2 replied:
done.

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+
+    @pytest.mark.run(order=9)
+    def test_create_route(self, apigw_validator, test_resource_values):
+        # Create an integration for the HTTP Api
```
s/integration/route/ :)

vijtrip2 replied:
Done.

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+        )
+        logging.error(f"route resource: {route_resource_data}")
+
+        # Create the k8s integration resource
```
s/integration/route/

vijtrip2 replied:
Done.

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+            "route",
+            additional_replacements=test_resource_values
+        )
+        logging.error(f"route resource: {route_resource_data}")
```
This should not be logging.error...

vijtrip2 replied:
changed to info

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+
+    @pytest.mark.run(order=10)
+    def test_create_stage(self, apigw_validator, test_resource_values):
+        # Create an integration for the HTTP Api
```
s/integration/stage/

vijtrip2 replied:
Done.

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+            "stage",
+            additional_replacements=test_resource_values
+        )
+        logging.error(f"stage resource: {stage_resource_data}")
```
logging.error...

vijtrip2 replied:
changed to info

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+        )
+        logging.error(f"stage resource: {stage_resource_data}")
+
+        # Create the k8s integration resource
```
s/integration/stage/

vijtrip2 replied:
Done

jaypipes commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,292 @@
+        response = requests.request(method='GET', url=invoke_url, headers={'Authorization': 'SecretToken'})
+        assert 200 == response.status_code
+
+    @pytest.mark.run(order=-5)
```
Did you realize that you're using negative numbers for the delete test orders?

RedbackThomson replied:
This is a genius way of telling pytest to run it in an order offset from the end of the tests. Matching the test that runs with `order=1` with its deletion test with `order=-1`, etc.

jaypipes reviewed (changes requested):
@vijtrip2 perhaps you forgot to push your fixes/changes?

RedbackThomson commented on test/e2e/apigatewayv2/tests/helper.py:
```diff
@@ -0,0 +1,209 @@
+STAGE_RESOURCE_PLURAL = 'stages'
+
+
+def api_ref_and_data(api_resource_name: str, replacement_values: dict):
```
For most of the resource loading methods in this directory, if they're being used for fixtures, you can probably re-use the same abstraction we created for the SageMaker resources.
We created a generic "load resource" method - https://github.com/aws-controllers-k8s/community/blob/main/test/e2e/common/k8s.py#L111
We then override it with the SageMaker defaults - https://github.com/aws-controllers-k8s/community/blob/main/test/e2e/sagemaker/__init__.py#L30

This should save time dealing with silly repetitive code.

vijtrip2 replied:
I will need to pull latest changes from main branch for this. I will try it tomorrow.

jaypipes replied:
@RedbackThomson I think @vijtrip2 can do that in a followup PR. I'd like to merge this PR soon and Vijay can improve it in later ones.

RedbackThomson commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,539 @@
+    assert k8s.get_resource_exists(api_ref)
+
+    api_id = cr['status']['apiID']
+    test_resource_values['API_ID'] = api_id
```
You're modifying the default replacement values? Might be worth putting `REPLACEMENT_VALUES.copy()` into this fixture, rather than globally. Just worried about the situation where multitasking the tests calls them out of order and you pass around modified values.

RedbackThomson replied:
Oh I think I see, you want to use the modified versions in all places? Why not just return a copy of the `REPLACEMENT_VALUES` as a return value from this fixture? Or just return the `api_id` as a single value?

RedbackThomson commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,539 @@
+        pytest.fail(f"expected {api_resource_name} to not exist. Did previous test cleanup?", False)
+    logging.debug(f"http api resource. name: {api_resource_name}, data: {api_data}")
+
+    k8s.create_custom_resource(api_ref, api_data)
+    cr = k8s.wait_resource_consumed_by_controller(api_ref)
+
+    assert cr is not None
+    assert k8s.get_resource_exists(api_ref)
```
Hopefully these unnecessary assertions are removed if you use the helpers I suggested above.

RedbackThomson commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,539 @@
+    yield api_ref, cr
+
+    k8s.delete_custom_resource(api_ref)
+    time.sleep(DELETE_WAIT_AFTER_SECONDS)
```
Is there a need to sleep after deleting the resource? Fixtures are cleaned up at the end of the test, so we don't care about waiting for it to be deleted.

`delete_custom_resource` actually returns an assertion that the resource was deleted. So if you want to be sure the CR is removed properly you can instead do:
```
_, deleted = k8s.delete_custom_resource(api_ref)
assert deleted
```

vijtrip2 replied:
good point. Removed the wait from fixtures.

RedbackThomson commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,539 @@
+@pytest.fixture(scope="module")
+def authorizer_resource(api_resource):
+    authorizer_resource_name = test_resource_values['AUTHORIZER_NAME']
+    authorizer_uri = "arn:aws:apigateway:{region}:lambda:path/2015-03-31/functions/{authorizer_function_arn}/invocations" \
```
nit: f-strings are the preferred method of string formatting

vijtrip2 replied:
Done.

RedbackThomson commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,539 @@
+    authorizer_ref, authorizer_data = helper.authorizer_ref_and_data(authorizer_resource_name=authorizer_resource_name,
+                                                                     replacement_values=test_resource_values)
+    if k8s.get_resource_exists(authorizer_ref):
+        pytest.fail(f"expected {authorizer_resource_name} to not exist. Did previous test cleanup?", False)
```
I had no idea you could even throw a `pytest.fail` inside a fixture. I think a `raise` would be more appropriate, as this isn't the test itself failing but the bootstrapping for said test.

vijtrip2 replied:
Agree. Done.

RedbackThomson commented on test/e2e/apigatewayv2/tests/test_api.py:
```diff
@@ -0,0 +1,539 @@
+class TestApiGatewayV2:
+
+    def test_perform_invocation(self, api_resource, stage_resource):
+        time.sleep(DEPLOYMENT_WAIT_AFTER_SECONDS)
```
What's the purpose of this sleep? We have already asserted, in the fixture ,that `k8s.wait_resource_consumed_by_controller`?

vijtrip2 replied:
> What's the purpose of this sleep? We have already asserted, in the fixture ,that `k8s.wait_resource_consumed_by_controller`?

This is APIGateway specific wait. Where we wait for changes to get propagated to apigateway dataplane before performing invocation.

jaypipes commented on test/e2e/apigatewayv2/service_bootstrap.py:
```diff
@@ -0,0 +1,100 @@
+from apigatewayv2.bootstrap_resources import TestBootstrapResources
+from common.aws import get_aws_region
+
+RAND_TEST_SUFFIX = (''.join(random.choice(string.ascii_lowercase) for _ in range(6)))
```
nit: outer parens not needed above. :)

jaypipes commented on test/e2e/apigatewayv2/service_bootstrap.py:
```diff
@@ -0,0 +1,100 @@
+    except lambda_client.exceptions.ResourceNotFoundException:
+        pass
+
+    with tempfile.TemporaryDirectory() as tempdir:
```
:+1:

jaypipes reviewed (approved):
I'm good with this. Certainly lots of places to remove duplicate code and clean things up, but those can be done in future PRs.

Thanks @vijtrip2!

A-Hilaly reviewed (approved):
/lgtm :+1:

vijtrip2 commented:
@RedbackThomson addressed all comments except load-resource one. Tests are passing and one minute faster. : )

```
➜  aws-controllers-k8s git:(pytest) ✗ make kind-test
checking AWS credentials ... ok.
creating kind cluster ack-test-8f357765-1c1e704d ... ok.
[+] Building 182.8s (21/21) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                                           0.1s
 => => transferring dockerfile: 39B                                                                                                                                                                                                            0.1s
 => [internal] load .dockerignore                                                                                                                                                                                                              0.1s
 => => transferring context: 2B                                                                                                                                                                                                                0.0s
 => [internal] load metadata for docker.io/library/golang:1.14.1                                                                                                                                                                               3.2s
 => [internal] load metadata for docker.io/library/amazonlinux:2                                                                                                                                                                               3.2s
 => [auth] library/amazonlinux:pull token for registry-1.docker.io                                                                                                                                                                             0.0s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                                                                                                                                  0.0s
 => CACHED [stage-1 1/3] FROM docker.io/library/amazonlinux:2@sha256:06380711d6a8ac0b6989f7e2a4419e560796791d9c7c843753a719c73552dc30                                                                                                          0.0s
 => [internal] load build context                                                                                                                                                                                                              0.1s
 => => transferring context: 8.71kB                                                                                                                                                                                                            0.1s
 => [builder  1/11] FROM docker.io/library/golang:1.14.1@sha256:38a520b05b971181a5a30d72b74d1b70c63688eb267a4fb376c7476a5c6f885b                                                                                                               0.0s
 => => resolve docker.io/library/golang:1.14.1@sha256:38a520b05b971181a5a30d72b74d1b70c63688eb267a4fb376c7476a5c6f885b                                                                                                                         0.0s
 => CACHED [builder  2/11] WORKDIR /github.com/aws-controllers-k8s/apigatewayv2-controller                                                                                                                                                     0.0s
 => CACHED [builder  3/11] COPY apigatewayv2-controller/LICENSE /github.com/aws-controllers-k8s/apigatewayv2-controller/LICENSE                                                                                                                0.0s
 => CACHED [builder  4/11] COPY apigatewayv2-controller/ATTRIBUTION.md /github.com/aws-controllers-k8s/apigatewayv2-controller/ATTRIBUTION.md                                                                                                  0.0s
 => CACHED [builder  5/11] COPY apigatewayv2-controller/go.mod /github.com/aws-controllers-k8s/apigatewayv2-controller/go.mod                                                                                                                  0.0s
 => CACHED [builder  6/11] COPY apigatewayv2-controller/go.sum /github.com/aws-controllers-k8s/apigatewayv2-controller/go.sum                                                                                                                  0.0s
 => [builder  7/11] RUN go mod download                                                                                                                                                                                                       51.8s
 => [builder  8/11] COPY apigatewayv2-controller/apis /github.com/aws-controllers-k8s/apigatewayv2-controller/apis                                                                                                                             0.1s
 => [builder  9/11] COPY apigatewayv2-controller/cmd /github.com/aws-controllers-k8s/apigatewayv2-controller/cmd                                                                                                                               0.1s
 => [builder 10/11] COPY apigatewayv2-controller/pkg /github.com/aws-controllers-k8s/apigatewayv2-controller/pkg                                                                                                                               0.1s
 => [builder 11/11] RUN GIT_VERSION=30ecf27 &&     GIT_COMMIT=30ecf27019adce0ced02728e8dd907f15741af2b &&     BUILD_DATE=2021-04-07T10:55 &&     go build -ldflags="-X github.com/aws-controllers-k8s/apigatewayv2-controller/pkg/version.G  124.7s
 => [stage-1 2/3] COPY --from=builder /github.com/aws-controllers-k8s/apigatewayv2-controller/bin/controller /github.com/aws-controllers-k8s/apigatewayv2-controller/LICENSE /github.com/aws-controllers-k8s/apigatewayv2-controller/ATTRIBUT  0.6s
 => exporting to image                                                                                                                                                                                                                         0.5s
 => => exporting layers                                                                                                                                                                                                                        0.5s
 => => writing image sha256:78fb218bd91f50c46ca91e5e4606182bf77fa8cf21b3c51241f7ba122ff55b5c                                                                                                                                                   0.0s
 => => naming to docker.io/library/aws-controllers-k8s:apigatewayv2-v0.0.2-69-g0b72768-dirty                                                                                                                                                   0.0s
ok.
loading the images into the cluster ... ok.
loading CRD manifests for apigatewayv2 into the cluster ... ok.
loading RBAC manifests for apigatewayv2 into the cluster ... ok.
loading service controller Deployment for apigatewayv2 into the cluster ...ok.
generating AWS temporary credentials and adding to env vars map ... ok.
======================================================================================================
To poke around your test cluster manually:
export KUBECONFIG=/Users/vijat/Documents/gocode/src/github.com/vijtrip2/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-8f357765-1c1e704d/kubeconfig
kubectl get pods -A
======================================================================================================
testing Helm release for apigatewayv2 for release version v0.0.2-69-g0b72768-dirty.
Building release artifacts for apigatewayv2-v0.0.2-69-g0b72768-dirty
Generating custom resource definitions for apigatewayv2
Generating RBAC manifests for apigatewayv2
namespace/ack-system-test-helm created
installing the helm chart for ack-apigatewayv2-controller in namespace ack-system-test-helm ... ok.
uninstalling the helm chart for ack-apigatewayv2-controller in namespace ack-system-test-helm ... ok.
namespace "ack-system-test-helm" deleted
INFO:root:Wrote bootstrap to /root/tests/apigatewayv2/bootstrap.yaml
=============================================================================================================== test session starts ================================================================================================================
platform linux -- Python 3.8.8, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /root/tests
plugins: xdist-2.2.0, forked-1.3.0
[gw0] Python 3.8.8 (default, Feb 19 2021, 18:07:06)  -- [GCC 8.3.0]
[gw1] Python 3.8.8 (default, Feb 19 2021, 18:07:06)  -- [GCC 8.3.0]
[gw2] Python 3.8.8 (default, Feb 19 2021, 18:07:06)  -- [GCC 8.3.0]
gw0 [7] / gw1 [7] / gw2 [7]
scheduling tests via LoadFileScheduling

apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_perform_invocation
[gw0] [ 14%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_perform_invocation
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_httpapi
[gw0] [ 28%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_httpapi
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_httpapi_using_import
[gw0] [ 42%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_httpapi_using_import
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_integration
[gw0] [ 57%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_integration
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_authorizer
[gw0] [ 71%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_authorizer
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_route
[gw0] [ 85%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_route
apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_stage
[gw0] [100%] PASSED apigatewayv2/tests/test_api.py::TestApiGatewayV2::test_crud_stage

========================================================================================================== 7 passed in 307.89s (0:05:07) ===========================================================================================================
To resume test with the same cluster use: " TMP_DIR=/Users/vijat/Documents/gocode/src/github.com/vijtrip2/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-8f357765-1c1e704d
    AWS_SERVICE_DOCKER_IMG=aws-controllers-k8s:apigatewayv2-v0.0.2-69-g0b72768-dirty "
```

RedbackThomson reviewed (approved):
LGTM. Other QoL changes can come later. Ship it!
